Calculate cargo price using a weighted average of acquisition prices

Tkael opened the pull request:
Fixes #1978.
Revises the cargo monitor to calculate a weighted average of the commander's acquisition price for each cargo item (rather than the galactic average price).
Fixes inaccurate projected profits for the `Commodity sales check` script.

Tkael committed:
Calculate cargo price using a weighted average of acquisition prices
Fixes #1978.

GioVAX commented on CargoMonitor/CargoMonitor.cs:
```diff
@@ -294,25 +300,24 @@ private bool _handleCommodityCollectedEvent(CommodityCollectedEvent @event)
             Cargo cargo = GetCargoWithEDName(@event.commodityDefinition?.edname);
             if (cargo != null)
             {
+                Haulage haulage = cargo.haulageData.FirstOrDefault(h => h.missionid == @event.missionid);
```
Can this be 'SingleOrDefault'?
Or can the collection contain multiple items with the same 'missionId'?

Tkael replied:
The general pattern Hooda appeared to follow was to try to look up haulage, then only add it if it was missing. Consequently, I don't think it's intended to contain multiple items with the same missionId but you're right that it isn't designed to prevent it. I have tried to write my changes to preclude that possibility, but since there is no duplicate protection built in I think that swapping to use 'SingleOrDefault' does create additional risk.

The same is true for cargo in the inventory collection.

We could consider creating a separate ticket to transition from lists to dictionaries for both the cargo in inventory and for haulage (using commodity `edname` as a unique key for cargo and haulage `missionid` as a unique key for haulage) but doing so could have implications for other dependent projects and is would have a wider impact than I intended for this first pass.

richardbuckle replied:
I would prefer to leave it as is for now.

GioVAX commented on CargoMonitor/CargoMonitor.cs:
```diff
@@ -493,28 +499,32 @@ private void _handleCargoDepotEvent(CargoDepotEvent @event)
                         {
                             // Cargo instantiated by either 'Mission accepted' event or previous 'WingUpdate' update
                             haulage = cargo.haulageData.FirstOrDefault(ha => ha.missionid == @event.missionid);
```
As above: 'SignleOrDefault'?

Tkael replied:
As above

GioVAX commented on CargoMonitor/CargoMonitor.cs:
```diff
@@ -404,21 +409,22 @@ private void handleCommodityPurchasedEvent(CommodityPurchasedEvent @event)
-            bool update = false;
-            Cargo cargo = GetCargoWithEDName(@event.commodityDefinition?.edname);
-            if (cargo != null)
+            Cargo cargo = GetCargoWithEDName(@event.commodityDefinition?.edname) ?? new Cargo(@event.commodityDefinition?.edname);
```
Can we move 'new Cargo(...)' inside 'GetCargoWithEDName'?
I see the same pattern below and it could remove some duplication, but I do not know if this is used in different ways in other parts of the code.

Tkael replied:
I count 6 references with methods which are conditional on whether the return value is null:
- _handleCommodityCollectedEvent
- _handleCommodityEjectedEvent
- _handleCommodityRefinedEvent
- handleCommoditySoldEvent
- _handleMissionCompletedEvent
- _handleEngineerContributedEvent

I also count 5 references where we instantiate `new Cargo` if the return is null. We could create a new method `GetOrCreateCargoWithEDName` to help distinguish this case if we think that doing so adds sufficient value.

richardbuckle replied:
Well, in general any "get by name" lookup *should* return null if the item is not there, and it is the caller's responsibility to decide what to do about it. If we did make this change, we would definitely have to call it `GetOrCreateCargoWithEDName` instead, otherwise the function name is lying to us.

But on reflection I think I rather prefer having the error handling out there and in our faces in this case.

GioVAX commented on DataDefinitions/Cargo.cs:
```diff
@@ -6,35 +6,34 @@
+        [Obsolete("Please use localizedName or invariantName")]
         public string name => localizedName;
 
+        [JsonProperty("edname")]
```
Nice  =)

GioVAX reviewed (commented):
Added some nits

Tkael committed:
Ensure that the `AddDetailedQty` method in Cargo.cs updates any existing haulage instance rather than generating a duplicate.
When we collect cargo, there is no need to pass haulage even if it is associated with a mission (the haulage is neither new nor updated).

richardbuckle commented on DataDefinitions/Cargo.cs:
```diff
@@ -142,8 +143,7 @@ private void OnDeserialized(StreamingContext context)
             {
                 // legacy JSON with no edname in the top level
-                string edname = (string)_additionalJsonData["commodity"]["edname"];
-                commodityDef = CommodityDefinition.FromEDName(edname);
```
I might be missing something here. Don't we need this to load the CommodityDefinition from JSON correctly?

Tkael replied:
Nope, it's redundant. What you're missing is this (from the same file):
```cs
        public string edname
        {
            get => commodityDef.edname;
            set
            {
                CommodityDefinition cDef = CommodityDefinition.FromEDName(value);
                this.commodityDef = cDef;
            }
        }
```

richardbuckle commented on DataDefinitions/Cargo.cs:
```diff
@@ -154,28 +154,115 @@ private void OnDeserialized(StreamingContext context)
+            if (amount == 0) { return; }
+            var weightedValueSum = (weightedAvgPrice * total) + (thisPrice * amount);
+            var weightedQtySum = total + amount;
+            weightedAvgPrice = weightedValueSum / weightedQtySum;
```
Do we need to guard against divide by zero here, such as when we sell everything?

Tkael replied:
Good call. Anytime we do division, we need to watch out for that.

richardbuckle commented on DataDefinitions/Cargo.cs:
```diff
@@ -154,28 +154,115 @@ private void OnDeserialized(StreamingContext context)
+        /// <param name="amount">The amount of cargo to add</param>
+        /// <param name="cargoHaulageData">Add or update haulage instance</param>
+        /// <param name="_price">The acquisition price per unit (if not zero)</param>
+        public void AddDetailedQty(CargoType cargoType, int amount, Haulage cargoHaulageData = null, int _price = 0)
```
A couple of things.

1. Leading underscores are a convention for private instance variables and methods. It's not conventional to use them to disambiguate a parameter from an existing instance variable and such a clash usually indicates that a better name is needed for the parameter. Suggest `acquistionPrice` here.

2. I'm not sure that defaulting `acquistionPrice` to zero is a good idea. I would rather make callers think about it and pass zero explicitly if appropriate. This would of course mean making it the 3rd param, but I am OK with that.

Tkael replied:
Good calls.

richardbuckle commented on DataDefinitions/Cargo.cs:
```diff
@@ -154,28 +154,115 @@ private void OnDeserialized(StreamingContext context)
+        }
+
+        public void CalculateWeightedPrice(int thisPrice, int amount)
+        {
```
"Calculate" implies no change to the object.
Let's rename this `UpdateWeightedPrice(int newPrice, int newAmount)`.

Tkael replied:
Agreed.

richardbuckle commented on DataDefinitions/Cargo.cs:
```diff
@@ -154,28 +154,115 @@ private void OnDeserialized(StreamingContext context)
+                                if (haulageData[i].missionid == cargoHaulageData.missionid)
+                                {
+                                    haulageData[i] = cargoHaulageData;
+                                    return;
```
A `return` this deep is always a bit of a loaded gun should the code be blindly moved into another context.

I would instead of the manual `for` loop use `haulageData.FindIndex()` with a predicate and update if found or Add() if not.

richardbuckle commented on DataDefinitions/Cargo.cs:
```diff
@@ -154,28 +154,115 @@ private void OnDeserialized(StreamingContext context)
             {
-                this.need = this.haulageData.Sum(h => h.need);
+                case CargoType.haulage:
+                    {
```
Such braces are only required if the block is introducing new local vars.

Tkael replied:
Fair. I think it's probably a good practice anyways.

richardbuckle commented on CargoMonitor/CargoMonitor.cs:
```diff
@@ -953,15 +1000,19 @@ private void readInventory(CargoMonitorConfiguration configuration = null)
+                    if (string.Equals(inventory[i].edname, cargo.edname, StringComparison.InvariantCultureIgnoreCase))
+                    {
+                        inventory[i] = cargo;
+                        return;
```
A `return` this deep is always a bit of a loaded gun should the code be blindly moved into another context.

I would instead of the manual `for` loop use `inventory.FindIndex()` with a predicate and update if found or Add() if not.

Tkael committed:
Fix Rollbar# 19464 - an exception is thrown when cAPI price data is not expressed as integers (https://rollbar.com/EDCD/EDDI/items/19464/)
```json
{
   "id": 128666747,
   "name": "LavianBrandy",
   "legality": "I",
   "buyPrice": 745.8000000000001,
   "sellPrice": 12833.203434782608,
   "meanPrice": 10365,
   "demandBracket": 0,
   "stockBracket": 0,
   "stock": 0,
   "demand": 0,
   "statusFlags": [],
   "categoryname": "Narcotics",
   "locName": "Lavian Brandy"
  }
```

richardbuckle committed:
Avoid return in AddOrUpdateCargo

richardbuckle committed:
Allow CargoMonitor ctor to accept and optional config param

richardbuckle committed:
Stop non-filebacked CargoMonitorConfigurations from writing to default location

richardbuckle committed:
`TestInitialize` methods **must** be `public` or they will not run.

richardbuckle committed:
Add JetBrains Annotations package

richardbuckle committed:
Restore no-param ctor and mark as PublicAPI

richardbuckle committed:
Configure the test cargoMonitor to be non-disk-based

richardbuckle committed:
Remove all file system calls from CargoMonitorTests`

Since the test class's config is no longer disk-based and cannot wriote to disk, it is no longer necessary to save and restore the actual live config.
Tested by adding a "canary" entry for my own config file.

richardbuckle committed:
Add JetBrains.Annotations to all

richardbuckle committed:
Amend failing `TestEddbGetSystem` test

Also removed unused `fakeEddbRestClient`. Not sure why it was introduced.

richardbuckle commented:
Created #1997 to encapsulate my desired fixes here.

Tkael committed:
Fix fixes to EDDB tests.

Tkael committed:
Tidy ups recommended by @richardbuckle
- rename methods and variables for clarity
- make acquisitionPrice a required rather than an optional parameter
- check for qty greater than zero before diving by qty
